test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/TempDataApplicationModelProviderTest.cs

```diff
@@ -0,0 +1,125 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using System.Reflection;
+using Microsoft.AspNetCore.Mvc.ApplicationModels;
+using Microsoft.AspNetCore.Mvc.Internal;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Mvc.ViewFeatures.Internal
+{
+    public class TempDataApplicationModelProviderTest
+    {
+        [Theory]
+        [InlineData(typeof(TestController_OneTempDataProperty))]
+        [InlineData(typeof(TestController_TwoTempDataProperties))]
+        public void AddsTempDataPropertyFilter_ForTempDataAttributeProperties(Type type)
+        {
+            // Arrange
+            var provider = new TempDataApplicationModelProvider();
+            var defaultProvider = new DefaultApplicationModelProvider(new TestOptionsManager<MvcOptions>());
+
+            var context = new ApplicationModelProviderContext(new[] { type.GetTypeInfo() });
+            defaultProvider.OnProvidersExecuting(context);
+
+            // Act
+            provider.OnProvidersExecuting(context);
+
+            // Assert
+            var controller = Assert.Single(context.Result.Controllers);
+            Assert.Single(controller.Filters, f => f is SaveTempDataPropertyFilterFactory);
+        }
+
+        [Fact]
+        public void InitializeFilterFactory_WithExpectedPropertyHelpers_ForTempDataAttributeProperties()
+        {
+            // Arrange
+            var provider = new TempDataApplicationModelProvider();
+            var defaultProvider = new DefaultApplicationModelProvider(new TestOptionsManager<MvcOptions>());
+
+            var context = new ApplicationModelProviderContext(new[] { typeof(TestController_OneTempDataProperty).GetTypeInfo() });
+            defaultProvider.OnProvidersExecuting(context);
+
+            // Act
+            provider.OnProvidersExecuting(context);
+            var controller = context.Result.Controllers.SingleOrDefault();
+            var filter = controller.Filters.OfType<SaveTempDataPropertyFilterFactory>();
+            var saveTempDataPropertyFilterFactory = filter.SingleOrDefault();
+
+            // Assert
+            Assert.NotNull(saveTempDataPropertyFilterFactory);
+            var tempDataPropertyHelper = Assert.Single(saveTempDataPropertyFilterFactory.TempDataProperties);
+            Assert.Equal("Test2", tempDataPropertyHelper.Name);
+            Assert.NotNull(tempDataPropertyHelper.Property.GetCustomAttribute<TempDataAttribute>());
```

Comment:
You could simply assert it's the exact property instead of verifying two pieces of information about it:

```
var expected = typeof(TestController_OneTempDataProperty).GetProperty(nameof(TestController_OneTempDataProperty.Test2);
...
Assert.Same(expected, tempDataPropertyHelper.Property);
```